presto-main/src/main/java/com/facebook/presto/sql/planner/plan/JoinNode.java

```diff
@@ -93,6 +97,59 @@ public JoinNode(@JsonProperty("id") PlanNodeId id,
         checkArgument(!isCrossJoin() || inputSymbols.equals(outputSymbols), "Cross join does not support output symbols pruning or reordering");
     }
 
+    public JoinNode flip()
+    {
+        return new JoinNode(
+                getId(),
+                flipType(type),
+                right,
+                left,
+                flipJoinCriteria(criteria),
+                flipOutputSymbols(getOutputSymbols(), left, right),
+                filter,
+                rightHashSymbol,
+                leftHashSymbol,
+                distributionType);
```

Comment:
I don't think it's necessary to add a check state (It doesn't mean anything to flip the distribution type, I think it's correct to keep it the same) , but I took the rest of your suggestion.